utils/tfw_logger.cc

```diff
@@ -0,0 +1,461 @@
+
+	auto col = std::make_shared<clickhouse::ColumnDateTime64>(3);
+	block->AppendColumn("timestamp", col);
+
+	for (i = TFW_MMAP_LOG_ADDR; i < TFW_MMAP_LOG_MAX; ++i) {
+		const TfwField *field = &tfw_fields[i];
+
+		auto col = tfw_column_factory(field->code);
+		block->AppendColumn(field->name, col);
+	}
+
+	return block;
+}
+
+int
+read_access_log_event(const char *data, int size, TfwClickhouse *clickhouse)
+{
+	auto block = clickhouse->get_block();
+	const char *p = data;
+	TfwBinLogEvent *event = (TfwBinLogEvent *)p;
+	int i;
+
+	p += sizeof(TfwBinLogEvent);
+	size -= sizeof(TfwBinLogEvent);
+
+#define INT_CASE(method, col_type, val_type)					\
+	case method:								\
+		if (TFW_MMAP_LOG_FIELD_IS_SET(event, i)) {			\
+			if (len > size)						\
+				return -1;					\
+			(*block)[ind]->As<col_type>()->Append(*(val_type *)p);	\
+		} else								\
+			(*block)[ind]->As<col_type>()->Append(0);		\
+		break;
+
+
+	(*block)[0]->As<clickhouse::ColumnDateTime64>()->Append(event->timestamp);
+
+	for (i = TFW_MMAP_LOG_ADDR; i < TFW_MMAP_LOG_MAX; ++i) {
+		int len, ind = i + 1;
+
+		len = tfw_mmap_log_field_len((TfwBinLogFields)i);
+
+		switch (i) {
+		INT_CASE(TFW_MMAP_LOG_ADDR,
+			 clickhouse::ColumnIPv6, struct in6_addr);
+		INT_CASE(TFW_MMAP_LOG_METHOD,
+			 clickhouse::ColumnUInt8, unsigned char);
+		INT_CASE(TFW_MMAP_LOG_VERSION,
+			 clickhouse::ColumnUInt8, unsigned char);
+		INT_CASE(TFW_MMAP_LOG_STATUS,
+			 clickhouse::ColumnUInt16, uint16_t);
+		INT_CASE(TFW_MMAP_LOG_RESP_CONT_LEN,
+			 clickhouse::ColumnUInt32, uint32_t);
+		INT_CASE(TFW_MMAP_LOG_RESP_TIME,
+			 clickhouse::ColumnUInt32, uint32_t);
+		INT_CASE(TFW_MMAP_LOG_DROPPED,
+			 clickhouse::ColumnUInt64, uint64_t);
+
+		case TFW_MMAP_LOG_VHOST:
+		case TFW_MMAP_LOG_URI:
+		case TFW_MMAP_LOG_REFERER:
+		case TFW_MMAP_LOG_USER_AGENT:
+			if (!TFW_MMAP_LOG_FIELD_IS_SET(event, i)) {
+				(*block)[ind]->As<clickhouse::ColumnString>()->Append(
+					std::move(std::string("")));
+				break;
+			}
+			len = *((uint16_t *)p);
+			if (len + 2 > size)
+				return -1;
+			(*block)[ind]->As<clickhouse::ColumnString>()->Append(
+				std::move(std::string(p + 2, len)));
+			len += 2;
+			break;
+		default:
+			throw Except("Unknown field type: {}", i);
+		}
+
+		if (TFW_MMAP_LOG_FIELD_IS_SET(event, i)) {
+			p += len;
+			size -= len;
+		}
+	}
+
+	return p - data;
+}
+
+void
+callback(const char *data, int size, void *private_data)
+{
+	TfwClickhouse *clickhouse = (TfwClickhouse *)private_data;
+	TfwBinLogEvent *event;
+	const char *p = data;
+	int r;
+
+	dbg_hexdump(data, size);
+
+	do {
+		if (size < (int)sizeof(TfwBinLogEvent))
+			return;
+
+		event = (TfwBinLogEvent *)p;
+
+		switch (event->type) {
+		case TFW_MMAP_LOG_TYPE_ACCESS:
+			r = read_access_log_event(p, size, clickhouse);
+			if (r < 0)
+				return;
+			size -= r;
+			p += r;
+			break;
+		default:
+			throw Except("Unsupported log type: {}",
+				     (unsigned int)event->type);
+			return;
+		}
+	} while (size > 0);
```

Comment:
This is an extra check for the check on line 213

Reply:
Fixed.